### ISSUE TITLE
Markdown links in hoverMessage don't correctly detect ending/right parenthesis

### ISSUE THREAD

**Issue by tgriff** [Microsoft/vscode#40172]
- VSCode Version: 1.18.1
- OS Version: Windows 10

I'm working on an VSCode extension that highlights lines and provides external links in the hover message.  Some of these links need to contain parentheses, but the markdown renderer for hover messages doesn't correctly detect interior parentheses.  This breaks these links.

Note that I don't see this issue in other places where markdown is rendered in VSCode (e.g. similar links in README.md render correctly).  I also noticed a similar issue got fixed in #26449 so I think that fix just needs to be applied anywhere else markdown links are being processed/rendered.

Here's a sample extension.ts that reproduces the issue:
```typescript
'use strict';
import * as vscode from 'vscode';

export function activate(context: vscode.ExtensionContext) {
    vscode.window.onDidChangeActiveTextEditor(switchedToFile, null, context.subscriptions);
}

function switchedToFile(textEditor: vscode.TextEditor) {
    const decoTypeRed = vscode.window.createTextEditorDecorationType({
        backgroundColor: 'rgba(255,0,0,0.4)'
    });

    // Links that contain an ending/right parenthesis ")" are broken.
    let markdown = "[This is a link](http://mysite.com?function=foo(int bar)&line=42)";

    let range = textEditor.document.lineAt(0).range;
    let decoOpts: vscode.DecorationOptions[] = [];
    decoOpts.push({
        range: range,
        hoverMessage: markdown
    });

    vscode.window.activeTextEditor.setDecorations(decoTypeRed, decoOpts);
}

export function deactivate() {
}
```
Here's the result:
![hovermessage_broken_link](https://user-images.githubusercontent.com/34492055/33953995-d0a38056-dfeb-11e7-9d1b-dfc6c47153bb.PNG)

**Comment by mjbvz** [comment]
Upstream issue with our core markdown parser: https://github.com/chjj/marked/issues/690

**Comment by michael-hawker** [comment]
Yup, another example with older ms doc links:

![image](https://user-images.githubusercontent.com/24302614/38693686-96c87210-3e3c-11e8-9a0c-3aa51b260490.png)

```
- [System.Windows.Media.RadialGradientBrush](https://msdn.microsoft.com/en-us/library/system.windows.media.radialgradientbrush(v=vs.110).aspx)
```

- [System.Windows.Media.RadialGradientBrush](https://msdn.microsoft.com/en-us/library/system.windows.media.radialgradientbrush(v=vs.110).aspx)

**Comment by mjbvz** [comment]
Fixed by 4daf8d2266a2b80a6fee6be1179f346beed3be73

For verification, use jsdocs:

```js
/**
 * [This is a link](http://example.com?function=foo(int%20bar)&line=42)
 * 
 * - [System.Windows.Media.RadialGradientBrush](https://msdn.microsoft.com/en-us/library/system.windows.media.radialgradientbrush(v=vs.110).aspx)
 */
const abc = ''
```

and hover over `abc` to make sure the markdown for the links is now parsed properly
